### PR TITLE
Generic request decoder and response encoder

### PR DESCRIPTION
**Pull request by adam-fowler**
Instead of requiring we store a request decoder and response encoder in the core context. This PR makes them accessible from the root of `HBRequestContext`. This allows us to make them to be generic and not existential types.

It does mean we have to get rid of `HBSetCodableMiddleware`, but I think making these generic is a bigger gain.

### CODE REVIEW

File: Sources/Hummingbird/Server/RequestContext.swift [hunk 1]
```diff
@@ -71,21 +61,21 @@ public struct HBCoreRequestContext: Sendable {
 /// Protocol that all request contexts should conform to. Holds data associated with
 /// a request. Provides context for request processing
 public protocol HBBaseRequestContext: Sendable {
+    associatedtype Decoder: HBRequestDecoder = NullDecoder
+    associatedtype Encoder: HBResponseEncoder = NullEncoder
```
Comment (Joannis) [inline Sources/Hummingbird/Server/RequestContext.swift:65]: Does it make sense to make these primary associated types?

Comment (Joannis) [inline Sources/Hummingbird/Server/RequestContext.swift:65]: This will need fixing in HBLambda

**Review by Joannis — APPROVED** [review]
One small possible improvement - if it's helpful

File: Sources/Hummingbird/Server/RequestContext.swift (hunk 1 above)
Comment (adam-fowler) [reply Sources/Hummingbird/Server/RequestContext.swift:65]: I'm not sure that is necessary. When are you going to need to specify the encoder in a parameter type

Comment (adam-fowler) [reply Sources/Hummingbird/Server/RequestContext.swift:65]: yep, PR on its way

**Pull request merged**